tests/test_crash_offset.py

```diff
@@ -0,0 +1,102 @@
+import os
+import subprocess
+import tempfile
+import contextlib
+import unittest
+
+import angr
+from angr.knowledge_plugins.key_definitions.constants import OP_BEFORE
+
+
+def crash_offset(program_path, function_call):
+    caller, callee, argument_index = function_call
+
+    project = angr.Project(program_path, auto_load_libs=False)
+    cfg = project.analyses.CFGFast(normalize=True, data_references=True)
+    project.analyses.CompleteCallingConventions(recover_variables=True, analyze_callsites=True)
+
+    caller = project.kb.functions[caller]
+    callee = project.kb.functions[callee]
+
+    argument_register = project.arch.registers[callee.arguments[argument_index].reg_name]
+    base_register = project.arch.registers[project.arch.register_names[project.arch.bp_offset]]
+
+    callee_node = cfg.get_any_node(callee.addr)
+    caller_node = next(iter(node for node in cfg.get_all_predecessors(callee_node) if node.name == caller.name))
+    call_address = list(caller_node.instruction_addrs)[-1]
+
+    rda = project.analyses.ReachingDefinitions(subject=caller, observation_points={("insn", call_address, OP_BEFORE)})
+    definitions = rda.get_reaching_definitions_by_insn(call_address, OP_BEFORE)
+    argument_value = definitions.registers.load(*argument_register).one_value()
+    base_value = definitions.registers.load(*base_register).one_value()
+
+    argument_offset = definitions.get_stack_offset(argument_value)
+    base_offset = definitions.get_stack_offset(base_value)
+
+    return base_offset - argument_offset + 8
+
+
+@contextlib.contextmanager
+def compiled_program(prefix_length, buffer_length, suffix_length):
+    program_template = """
+    #include <stdio.h>
+    #include <unistd.h>
+
+    int main()
+    {{
+        struct {{
+            char prefix[{prefix_length}];
+            char buffer[{buffer_length}];
+            char suffix[{suffix_length}];
+        }} stack;
+        read(0, stack.buffer, {prefix_length} + {buffer_length} + {suffix_length} + 0x1000);
+        return 0;
+    }}
+    """
+
+    program_code = program_template.format(
+        prefix_length=prefix_length, buffer_length=buffer_length, suffix_length=suffix_length
+    )
+
+    with tempfile.NamedTemporaryFile(suffix=".c", delete=False) as f:
+        program_path = f.name
+        f.write(program_code.encode())
+
+    binary_path = tempfile.mktemp()
+
+    subprocess.run(["gcc", program_path, "-o", binary_path, "-fno-stack-protector"])
```

Comment:
Is this test only intended to run on linux? This is gonna fail nightlies on macos and windows

Reply:
It honestly may be a better idea to pre-compile binaries and put them in the binaries repo rather than compiling this every time the test is run, especially since nowhere does angr explicitly depend on gcc,

Reply:
Why is this going to fail on macos/windows?

Reply:
Would you really rather ship a bunch of precompiled binaries for this? The very small source code being included here makes the testcase much more readable.

Can we make angr testcases explicitly depend on gcc? I don't think angr needs to, just angr testcases.

Reply:
I suppose gcc is aliased to clang by default on mac, but on windows it is not standard. I would not expect gcc to be in a user's path on windows. angr users msvc on windows.

Reply:
The prevailing wisdom on the angr team is that it's better to ship binaries so we don't have to deal with testcase failures when gcc updates :)